### ISSUE TITLE
graphql plugin severely degrades performance

### ISSUE THREAD

**Issue by andrew-kolesnikov** [DataDog/dd-trace-js#2109]
When `graphql` plugin is enabled, there is a noticeable drop in performance of the application.

To illustrate, I measured an **expensive** gql query with 10KB request payload, 1.6MB response, generating 450+ spans - most spans are _**graphql.resolve**_

| Description                                                | Measurement                   |
| ---------------------------------------------------------- | ----------------------------- |
| w/o trace                                                  | 168.065ms
| w/o plugins                                                | 166.232ms (no impact)
| w/trace defaults                                           | 212.490ms
| w/graphql only                                             | 225.042ms

**Environment**
isolated bench

* **Node.js version:**
node v16.13 LTS
* **Tracer version:**
latest dd-trace@2.8.0

Relevant dependencies:
    "@apollo/subgraph": "^0.1.2",
    "apollo-datasource": "^3.3.1",
    "apollo-graphql": "^0.9.5",
    "apollo-server-cache-redis": "^3.3.1",
    "apollo-server-caching": "^3.3.0",
    "apollo-server-core": "^3.1.1",
    "apollo-server-express": "^3.1.1",
    "apollo-server-plugin-response-cache": "^3.6.0",
    "dataloader": "^2.0.0",
    "datasource-sql": "^1.6.0",
    "dd-trace": "^2.8.0",
    "easygraphql-tester": "^6.0.1",
    "express": "^4.17.1",
    "graphql": "^15.7.2",
    "graphql-fields": "^2.0.3",
    "graphql-jit": "^0.7.4",
    "graphql-middleware": "^6.1.7",

**Comment by andrew-kolesnikov** [comment]
Another issue is that setting graphql `depth` seems to further degrade performance:
| Description                                                | Measurement                   |
| ---------------------------------------------------------- | ----------------------------- |
| w/o trace                                                  | 168.065ms
| w/o plugins                                                | 166.232ms (no impact)
| w/trace defaults                                           | 212.490ms
| w/graphql depth0                                           | 286.779ms
| w/graphql depth3                                           | 278.366ms
| w/graphql depth3, no signature                             | 272.034ms
| w/graphql, depth3, collapse                                | 276.342ms
| w/graphql only                                             | 225.042ms

**Comment by andrew-kolesnikov** [comment]
Another observation is that for **small response payloads** the issue is not as pronounced:
| Description                                                | Measurement                   |
| ---------------------------------------------------------- | ----------------------------- |
| w/o tracer                                                 |   45.301ms                    | 
| w/tracer defaults                                          |   47.905ms                    |

**Comment by rochdev** [comment]
We're in the process of rewriting this integration, which will hopefully help with performance. However, we do wrap every single resolver right now even with a depth set to 0, which has a cost. We could optimize that so that we don't patch at all for that case specifically which could help reduce the overhead when configured.

We've also seen cases where the overhead came from `async_hooks` and not from the tracer, so we should validate that. Can you try removing the tracer completely and adding an empty async hook like this:

```js
require('async_hooks').createHook({
  init () {}
}).enable()
```

If that has similar overhead then we'll know that this is what we need to work on, otherwise it would really be the integration itself.

**Comment by andrew-kolesnikov** [comment]
Measured as @rochdev suggested above, results indicate the issue is **not caused by async_hooks**:

| Small response                                                | Measurement                   |
| ---------------------------------------------------------- | ----------------------------- |
| w/o trace w/o async_hooks                            |   45ms                        | 
| w/o trace, async_hooks                               |   46ms                        | 
| default trace                                        |   46ms                        | 

| Large response                                                | Measurement                   |
| ---------------------------------------------------------- | ----------------------------- |
| w/o trace w/o async_hooks                            |   155ms                       | 
| w/o trace, async_hooks                               |   156ms                       | 
| default trace                                        |   **205ms**                       |

**Comment by rochdev** [comment]
@andrew-kolesnikov Thanks for confirming. Since we are re-writing the plugin anyway, I would say to try the new version when it lands, and then we can go from there.

**Comment by rochdev** [comment]
The new plugin is available in the latest version of dd-trace. It seems to have similar performance issues that we'll be looking into, however we have now optimized the plugin to avoid the additional overhead of instrumenting resolvers when `depth: 0` is set, so for now I would recommend using that which should remove the overhead almost completely as only the execution of the entire query will be traced.

**Comment by andrew-kolesnikov** [comment]
Thank you @rochdev for working on this and for posting an update - we have a lot of micro-services that are going to benefit from the reduction in overhead.

**Comment by bneigher** [comment]
hey @rochdev.  I tried setting the plugin init like this:

```javascript
import tracer from 'dd-trace'

// initialized in a different file to avoid hoisting.
tracer.init({
  logInjection: true
})
tracer.use('graphql', {
  depth: 0
})

export default tracer
```

but it is still super slow (on 2.10.0). it's fine on 2.8.0.  Am I setting depth: 0 incorrectly?  I'm on apollo/graphql/nestjs

**Comment by rochdev** [comment]
@bneigher Just to be sure that the option works, are you still getting `graphql.resolve` spans or did they at least stop coming in?

**Comment by rochdev** [comment]
Closing since this has been resolved in 2.11.0. We'll be making further improvements to the GraphQL integration but the regression described in this issue is fixed.

**Comment by bneigher** [comment]
thanks fellas
